src/section/section_keys.rs

```diff
@@ -50,12 +65,92 @@ impl SectionKeysProvider {
 
     pub fn finalise_dkg(&mut self, public_key: &bls::PublicKey) {
         if let Some(share) = &self.pending {
-            let pending_public_key = share.public_key_set.public_key();
+            if *public_key != share.public_key_set.public_key() {
+                return;
+            }
+        }
+        if let Some(share) = self.pending.take() {
+            if let Some(evicted) = self.cache.add(public_key, share) {
+                trace!("evicted old key from cache: {:?}", evicted);
+            }
+            trace!("finalised DKG: {:?}", public_key);
+        }
+    }
+}
+
+/// Implementation of simple cache.
+#[derive(Debug)]
+pub struct KeyCache {
+    map: BTreeMap<Vec<u8>, SectionKeyShare>,
+    list: VecDeque<bls::PublicKey>,
+    capacity: usize,
+}
+
+impl KeyCache {
+    /// Constructor for capacity based `KeyCache`.
+    pub fn with_capacity(capacity: usize) -> KeyCache {
+        KeyCache {
+            map: BTreeMap::new(),
+            list: VecDeque::with_capacity(capacity),
+            capacity,
+        }
+    }
 
-            if pending_public_key == *public_key {
-                trace!("finalise DKG: {:?}", pending_public_key);
-                self.current = self.pending.take();
+    ///
+    pub fn has_key_share(&self) -> bool {
+        !self.list.is_empty()
+    }
+
+    ///
+    pub fn get_most_recent(&self) -> Result<&SectionKeyShare> {
+        if !self.list.is_empty() {
+            if let Some(public_key) = self.list.get(self.list.len() - 1) {
+                if let Some(share) = self.map.get(&bincode::serialize(public_key)?) {
```

Comment:
`bls::PublicKey` already implements `Ord` so there should be no need to serialize it.

Reply:
Wasn't accepted as key though, needed Eq.